graphql/query_resolver.go

```diff
@@ -968,21 +968,48 @@ func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions)
 
 	opts := model.WaterfallOptions{
 		Limit:      limit,
-		Requesters: requesters,
 		MaxOrder:   maxOrderOpt,
 		MinOrder:   minOrderOpt,
-		Variants:   options.Variants,
+		Requesters: requesters,
+		Statuses:   utility.FilterSlice(options.Statuses, func(s string) bool { return s != "" }),
+		Tasks:      utility.FilterSlice(options.Tasks, func(s string) bool { return s != "" }),
+		Variants:   utility.FilterSlice(options.Variants, func(s string) bool { return s != "" }),
 	}
 
-	activeVersions, err := model.GetActiveWaterfallVersions(ctx, projectId, opts)
+	mostRecentWaterfallVersion, err := model.GetMostRecentWaterfallVersion(ctx, projectId)
 	if err != nil {
-		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting active waterfall versions: %s", err.Error()))
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("fetching most recent waterfall version: %s", err.Error()))
+	}
+	if mostRecentWaterfallVersion == nil {
+		return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("no versions found for project '%s'", projectId))
+	}
+
+	var activeVersions []model.Version
+	if len(opts.Tasks) > 0 || len(opts.Statuses) > 0 {
+		var searchOffset int
+		if opts.MaxOrder != 0 {
+			searchOffset = opts.MaxOrder
+		} else if opts.MinOrder != 0 {
+			searchOffset = opts.MinOrder
+		} else {
+			// Add one because minOrder and maxOrder are exclusive
+			searchOffset = mostRecentWaterfallVersion.RevisionOrderNumber + 1
+		}
+		activeVersions, err = model.GetActiveVersionsByTaskFilters(ctx, projectId, opts, searchOffset)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting active waterfall versions: %s", err.Error()))
+		}
+	} else {
+		activeVersions, err = model.GetActiveWaterfallVersions(ctx, projectId, opts)
```

Comment:
could the `searchOffset` be passed down to the build variant pipeline as well or is that something different?

Reply:
Yeah, I'll implement that in #8808 to avoid scope creep.